File: hackerrank/problem solving/algorithms/2 - implementation/21 - utopian tree/utopian_tree.java
// https://www.hackerrank.com/challenges/utopian-tree/problem?isFullScreen=true
// Java 25

void main() {
    int n = Integer.parseInt(IO.readln());
    int[] utopianTree = generateHeightsUtopianTree();
    for (int i = 0; i < n; i++) {
        int cycles = Integer.parseInt(IO.readln());
        IO.println(utopianTree[cycles]);
    }
}

// k: length of output array heights = 61
// T: O(k) = O(61) = O(1). S: O(61) = O(1) extra space.
int[] generateHeightsUtopianTree() {
    final int maxCycles = 60;
    int[] heights = new int[maxCycles + 1];

    for (int cycle = 0, n = maxCycles + 1, height = 0; cycle < n; cycle++) {
        height += isCycleHappeningDuringSpring(cycle) ? height : 1;
        heights[cycle] = height;
    }

    return heights;
}

boolean isCycleHappeningDuringSpring(int cycle) {
    return (cycle & 1) == 1;
}

File: hackerrank/problem solving/algorithms/2 - implementation/21 - utopian tree/utopian_tree.cpp
// https://www.hackerrank.com/challenges/utopian-tree/problem?isFullScreen=true
// C++23

import std;
using namespace std;

vector<int> generate_heights_utopian_tree();
bool is_cycle_happening_during_spring(int cycle);

int main()
{
    int n;
    cin >> n;
    vector utopian_tree{ generate_heights_utopian_tree() };
    for (int i{}; i < n; ++i) {
        int cycles;
        cin >> cycles;
        println("{}", utopian_tree[cycles]);
    }
    return 0;
}

// k: length of output array heights = 61
// T: O(k) = O(61) = O(1). S: O(61) = O(1) extra space.
vector<int> generate_heights_utopian_tree()
{
    constexpr int max_cycles{ 60 };
    vector<int> heights(max_cycles + 1);
    for (int height{}; auto [cycle, x] : views::enumerate(heights))
        x = height += is_cycle_happening_during_spring(cycle) ? height : 1;
    return heights;
}

bool is_cycle_happening_during_spring(int cycle)
{
    return cycle & 1;
}
